### PR TITLE
[GQSP] hamiltonian simulation

### PR DESCRIPTION
**Pull request by anurudhp**
on top of https://github.com/quantumlib/Qualtran/pull/848

### CODE REVIEW

File: qualtran/bloqs/basic_gates/su2_rotation.py [hunk 1]
```diff
@@ -144,6 +146,16 @@ def _t_complexity_(self) -> TComplexity:
     def _is_parameterized_(self) -> bool:
         return cirq.is_parameterized((self.theta, self.phi, self.lambd, self.global_shift))
 
+    @staticmethod
```
Comment (tanujkhattar) [inline qualtran/bloqs/basic_gates/su2_rotation.py:149]: Use a `@classmethod` instead of a static method.

File: qualtran/bloqs/hamiltonian_simulation/hamiltonian_simulation_by_gqsp.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            Motlagh and Wiebe. (2023). Theorem 7.
+    """
+
+    walk_operator: QubitizationWalkOperator
+    t: SymbolicFloat = field(kw_only=True)
+    alpha: SymbolicFloat = field(kw_only=True)
+    precision: SymbolicFloat = field(kw_only=True)
+
+    def _parameterized_(self):
+        return cirq.is_parameterized((self.t, self.alpha, self.precision))
```
Comment (tanujkhattar) [inline qualtran/bloqs/hamiltonian_simulation/hamiltonian_simulation_by_gqsp.py:64]: You can also use `is_symbolic(self.t, self.alpha, self.precision)` where `is_symbolic` is defined in qualtran/resource_counting/symbolic_counting_utils.py; to reduce the dependence on cirq

File: qualtran/bloqs/qsp/polynomial_approximations.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+    Args:
+        t: scale in the exponent
+        degree: value of $d$ to truncate the polynomial to $n \in [-d, d]$
+    """
+    coeff_indices = np.arange(-degree, degree + 1)
+    return 1j**coeff_indices * scipy.special.jv(coeff_indices, t)
+
+
+def approx_exp_sin_by_jacobi_anger(t: float, *, degree: int):
```
Comment (tanujkhattar) [inline qualtran/bloqs/qsp/polynomial_approximations.py:71]: Add return type annotations to all functions here

File: qualtran/bloqs/qsp/polynomial_approximations.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import bisect
+
+import numpy as np
+import scipy
+
+
+def degree_jacobi_anger_approximation(t: float, *, precision: float) -> int:
```
Comment (tanujkhattar) [inline qualtran/bloqs/qsp/polynomial_approximations.py:20]: Let's maybe modify the return type so that we return the symbolic expression encoded in `HamiltonianSimulationByGQSP.degree` if any of `t` or `precision` is symbolic here? Then the hamiltonian simulation class can always delegate to `degree_jacobi_anger_approximation` to find the degree and the symbolic expression can be used outside the context of hamiltonian simulation as well. 

There is nothing special about the symbolic expression that makes it specific to the Hamiltonian simulation case; correct?

File: qualtran/bloqs/qsp/polynomial_approximations.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
+#  Copyright 2024 Google LLC
```
Comment (tanujkhattar) [inline qualtran/bloqs/qsp/polynomial_approximations.py:1]: I think we should move this file to `qualtran/linalg/` directory instead of having it here in `qualtran/bloqs/qsp/`. The reason is that I imagine we'll be adding more such linalg utility methods that can be used to approximate functions as polynomials. One example we already have is `get_inverse_square_root_poly_coeffs` which approximates the inverse sqrt-root function as a polynomial which is the used in Bloqs for function approximations using QROM. 

Tl;Dr - Please move the file to `qualtran/linalg/` and maybe give a more specific name like `approx_exp_by_jacobi_anger.py`

File: qualtran/bloqs/hamiltonian_simulation/hamiltonian_simulation_by_gqsp.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+            }
+            prepare_out_soqs = bb.add_d(
+                prepare.adjoint() if adjoint else prepare,
+                **selection_registers,
+                **state_prep_ancilla_soqs,
+            )
+            gqsp_soqs |= {
+                reg.name: prepare_out_soqs.pop(reg.name) for reg in prepare.selection_registers
+            }
+            return gqsp_soqs, prepare_out_soqs
```
Comment (tanujkhattar) [inline qualtran/bloqs/hamiltonian_simulation/hamiltonian_simulation_by_gqsp.py:109]: Can this be a private method on the class instead of a nested local function?

**Review by tanujkhattar — COMMENTED** [review]
Another round of comments. Looking really good!

File: qualtran/bloqs/qsp/polynomial_approximations.py (hunk 4 above)
Comment (anurudhp) [reply qualtran/bloqs/qsp/polynomial_approximations.py]: yep, does not depend on the use case. moved it now.

File: qualtran/bloqs/qsp/polynomial_approximations.py (hunk 5 above)
Comment (anurudhp) [reply qualtran/bloqs/qsp/polynomial_approximations.py]: moved

File: qualtran/bloqs/for_testing/random_select_and_prepare.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+@frozen
+class RandomPrepareOracle(PrepareOracle):
+    U: RandomGate
+
+    @property
+    def selection_registers(self) -> tuple[Register, ...]:
+        return (Register('selection', BoundedQUInt(bitsize=self.U.bitsize)),)
+
+    @staticmethod
```
Comment (tanujkhattar) [inline qualtran/bloqs/for_testing/random_select_and_prepare.py:37]: nit: Use classmethod instead of staticmethod.

File: qualtran/bloqs/for_testing/random_select_and_prepare.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+from qualtran import BloqBuilder, BoundedQUInt, QBit, Register, SoquetT
+from qualtran.bloqs.for_testing.random_gate import RandomGate
+from qualtran.bloqs.qubitization_walk_operator import QubitizationWalkOperator
+from qualtran.bloqs.select_and_prepare import PrepareOracle, SelectOracle
+
+
+@frozen
+class RandomPrepareOracle(PrepareOracle):
+    U: RandomGate
```
Comment (tanujkhattar) [inline qualtran/bloqs/for_testing/random_select_and_prepare.py:31]: Add a docstring to the Bloq. Should we add a post init check to verify that first column of the matrix is all real numbers?

File: qualtran/bloqs/for_testing/random_select_and_prepare.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+    @cached_property
+    def alphas(self):
+        column = np.array(self.U.matrix)[:, 0]
+        np.testing.assert_almost_equal(np.imag(column), 0)
+        return column**2
+
+
+@frozen
+class PauliSelectOracle(SelectOracle):
```
Comment (tanujkhattar) [inline qualtran/bloqs/for_testing/random_select_and_prepare.py:69]: Please add a docstring to the class. Specify things like number of qubits it needs and whether it allocates additional qubits as part of the decomposition.

File: qualtran/bloqs/for_testing/random_select_and_prepare.py [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+        for cv, U in enumerate(self.select_unitaries):
+            bits = tuple(map(int, bin(cv)[2:].zfill(self.select_bitsize)))[::-1]
+            if self.control_val is not None:
+                bits = (*bits, self.control_val)
+            yield U.on(*target).controlled_by(*selection, control_values=bits)
+
+
+def random_qubitization_walk_operator(
+    select_bitsize: int, target_bitsize: int, *, random_state: np.random.RandomState
+) -> tuple[QubitizationWalkOperator, cirq.PauliSum]:
```
Comment (tanujkhattar) [inline qualtran/bloqs/for_testing/random_select_and_prepare.py:144]: Please add a docstring

File: qualtran/bloqs/hamiltonian_simulation/hamiltonian_simulation_by_gqsp.py [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+        r"""degree of the polynomial to approximate the function e^{it\cos(\theta)}"""
+        return degree_jacobi_anger_approximation(self.t * self.alpha, precision=self.precision)
+
+    @cached_property
+    def approx_cos(self) -> NDArray[np.complex_]:
+        r"""polynomial approximation for $$e^{i\theta} \mapsto e^{it\cos(\theta)}$$"""
+        if self._parameterized_():
+            raise ValueError(f"cannot compute `cos` approximation for parameterized Bloq {self}")
+        poly = approx_exp_cos_by_jacobi_anger(-self.t * self.alpha, degree=self.degree)
+        poly = scale_down_to_qsp_polynomial(poly) * (1 - 2 * self.precision)
```
Comment (tanujkhattar) [inline qualtran/bloqs/hamiltonian_simulation/hamiltonian_simulation_by_gqsp.py:115]: Maybe expand the docstring to explain why the scale down is needed and why we are multiplying by `(1 - 2 * eps)`

File: qualtran/bloqs/hamiltonian_simulation/hamiltonian_simulation_by_gqsp.py [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+        gqsp_soqs |= {
+            reg.name: prepare_out_soqs.pop(reg.name) for reg in prepare.selection_registers
+        }
+        return gqsp_soqs, prepare_out_soqs
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', **soqs: 'SoquetT') -> Dict[str, 'SoquetT']:
+        state_prep_ancilla = {
+            reg.name: bb.allocate(reg.total_bits())
+            for reg in self.walk_operator.prepare.junk_registers
+        }
```
Comment (tanujkhattar) [inline qualtran/bloqs/hamiltonian_simulation/hamiltonian_simulation_by_gqsp.py:157]: I think this wouldn't work with the cirq style API but we can fix it later.

`Allocate` and `Free` bloqs need to implement `as_cirq_op`. The reason it works right now is probably because there are no junk registers in the tested bloq so no new qubits are being allocated as part of the decomposition. Let's open an issue with a minimal example to reproduce.

cc @skushnir123 has run into this as well.

File: qualtran/bloqs/qsp/generalized_qsp.py [hunk 13]
```diff
@@ -214,6 +202,44 @@ def safe_angle(x):
...
+
+    points = np.exp(2j * np.pi * np.linspace(0, 1, num=n_points))
+    max_value = np.max(np.abs(Polynomial(P)(points)))
+    if max_value > 1:
+        P = P / max_value
+
+    return P
+
+
+def assert_is_qsp_polynomial(P, *, rtol: float, n_points: int = 100000):
```
Comment (tanujkhattar) [inline qualtran/bloqs/qsp/generalized_qsp.py:230]: Add type annotations to `P`.

File: qualtran/bloqs/qsp/generalized_qsp_test.py [hunk 14]
```diff
@@ -283,3 +284,29 @@ def test_generalized_real_qsp_with_symbolic_signal_matrix(degree: int):
...
+        approx_exp_cos_by_jacobi_anger,
+        degree_jacobi_anger_approximation,
+    )
+
+    random_state = np.random.RandomState(42 + int(t))
+
+    d = degree_jacobi_anger_approximation(t, precision=precision)
+
+    P = approx_exp_cos_by_jacobi_anger(t, degree=d)
+    # TODO compute optimal scaling factor
```
Comment (tanujkhattar) [inline qualtran/bloqs/qsp/generalized_qsp_test.py:302]: Add a link to the relevant open issue for this TODO

File: qualtran/bloqs/qsp/generalized_qsp.py [hunk 15]
```diff
@@ -214,6 +202,44 @@ def safe_angle(x):
...
+def scale_down_to_qsp_polynomial(
+    P: Sequence[complex], *, n_points: int = 10**5
+) -> NDArray[np.complex_]:
+    r"""Scale down the polynomial to be a valid QSP Polynomial
+
+    $P$ is a QSP polynomial if $|P(e^{i\theta})| \le 1$ for every $\theta \in [0, 2\pi]$.
+
+    If the input polynomial is not a valid QSP polynomial, this function attempts to compute
+    the maximum absolute value on the unit circle, and scale it down by that factor.
+    Otherwise returns the input as-is.
```
Comment (tanujkhattar) [inline qualtran/bloqs/qsp/generalized_qsp.py:214]: Add a link to the open issue where we are discussing potential improvements to this method.

**Review by tanujkhattar — APPROVED** [review]
LGTM overall. I've left a few comments mainly around improving documentation. Thanks for the amazing work getting this PR in shape!

File: qualtran/bloqs/for_testing/random_select_and_prepare.py (hunk 8 above)
Comment (anurudhp) [reply qualtran/bloqs/for_testing/random_select_and_prepare.py:48]: yep! added a post init check

File: qualtran/bloqs/hamiltonian_simulation/hamiltonian_simulation_by_gqsp.py (hunk 11 above)
Comment (anurudhp) [reply qualtran/bloqs/hamiltonian_simulation/hamiltonian_simulation_by_gqsp.py]: added a comment, and linked to #860.

File: qualtran/bloqs/hamiltonian_simulation/hamiltonian_simulation_by_gqsp.py (hunk 12 above)
Comment (anurudhp) [reply qualtran/bloqs/hamiltonian_simulation/hamiltonian_simulation_by_gqsp.py:159]: for the example, which cirq function should I test against? Because `cirq.unitary` doesn't make sense if there are junk registers (or does it?).

**Review by tanujkhattar — APPROVED** [review]

File: qualtran/bloqs/hamiltonian_simulation/hamiltonian_simulation_by_gqsp.py (hunk 12 above)
Comment (anurudhp) [reply qualtran/bloqs/hamiltonian_simulation/hamiltonian_simulation_by_gqsp.py:159]: Using `allocate` and `free` does seem to work with cirq?
```py
from qualtran import BloqBuilder
from qualtran.bloqs.basic_gates import Hadamard

bb = BloqBuilder()
q = bb.allocate(1)
q = bb.add(Hadamard(), q=q)
bb.free(q)
bloq = bb.finalize()

bloq.to_cirq_circuit()
```

```
(_c(0): ───alloc───H───free───, {})
```

Comment (tanujkhattar) [reply qualtran/bloqs/hamiltonian_simulation/hamiltonian_simulation_by_gqsp.py:159]: you can test against generating the cirq circuit for the decomposition. So `cirq.Circuit(cirq.decompose_once(GateHelper(bloq).operation))`

Comment (anurudhp) [reply qualtran/bloqs/hamiltonian_simulation/hamiltonian_simulation_by_gqsp.py:159]: I made a simple bloq which allocates and frees a qubit. The cirq conversion seems to work as expected.

```py
import cirq
from qualtran import BloqBuilder, GateWithRegisters, Signature
from qualtran.bloqs.basic_gates import CNOT, TGate
from qualtran.cirq_interop.testing import GateHelper


class MyBloq(GateWithRegisters):
    @property
    def signature(self) -> Signature:
        return Signature.build(q=1)

    def build_composite_bloq(self, bb: 'BloqBuilder', q) -> dict[str, 'SoquetT']:
        a = bb.allocate(1)
        q, a = bb.add(CNOT(), ctrl=q, target=a)
        q = bb.add(TGate(), q=q)
        bb.free(a)
        return {'q': q}


print(cirq.Circuit(cirq.decompose_once(GateHelper(MyBloq()).operation)))
```

ouputs:
```
_decompose_protocol_0_c(0): ───alloc───X───free───
                                       │
q: ────────────────────────────────────@───T──────
```
